rfcs/002-merge-types-terms-syntax.md

```diff
@@ -0,0 +1,287 @@
+available, otherwise, we would have removed it too).
+
+#### Records
+
+We need to allow mixing row types syntax and normal types syntax, that is `{foo
+: Bar, baz = 1, ..}`, or the parser will have a hard time disambiguating e.g.
+`{foo : Num, bar : Num, foo = 1, bar = 1}` (a term with piecewise signatures)
+from the previously record type `{foo: Num, bar: Num}`. This requires to resolve
+the conflict about `|`: we propose to use `;` to denote the record tail instead:
+`{foo : Num; a}`. We could also use a `..a` syntax: it may clash with potential
+uses for record extension as in say Rust: `Struct {bar: 1, ..rest}`, but we
+would probably use merging in Nickel for that.
+
+The `{_: Num}` syntax may look ambiguous, but this one can be special cased, as
+it is the case within the current type syntax (it only requires a one character
+look-ahead to decide).
+
+#### Enums
+
+With the first release in mind, which is the scope of this proposal, we propose
+to just disable support for enum altogether. This is a handy but hardly
+fundamental feature, and this lets us more time to find a good syntax
+replacement.
+
+### Translation
+
+Let's now define the `term` and `type` translations. Our design principle is to
+try as hard as possible to have them just be simple syntactic translations that
+insert the implicit `#` and `§` appropriately, if possible without having to
+recurse into terms or do something complex, although that may not be always
+possible.
+
+The two functions must satisfies the following coherence laws with `!` the
+embedding of the current syntax to the unified syntax (it may have to do some
+changes, like chainge `|` to `;`, but can surely easily define an embedding):
+`!: Term + Type -> UniTerm` and `~` is some flavour of operational equivalence
+(we won't prove those rules formally, as this a mere design guideline, so we
+also let the definitions be rough). It basically just erases `#` and `§`.
+
+1. For any `t: Term`, `term(!t) ~ t` and for any `T: Type`, `type(!T) = T` .
+   That is, a term or a type in the current syntax with translation operator
+   removed will be interpreted the same in the new syntax.
+2. For any `T: Type`, `term(!T) ~ §T`. That is, the interpretation as a term of
+   something that is a type in the current syntax must be the corresponding
+   contract.
+
+### `term`
+
+As explained, we see the `term` translation as inserting implicit `§`
+automatically.
+
+- For most term-only constructs (that is everything excepted
+  records and constructors coming from types), `term` is an homomorphism (e.g.
+  for lists `term [t1, ..., tn] = [ term(t1), ..., term(tn)]`).
+- For type-only constructs (ground types, `->`, and `{_: T}`), it is
+  `term(T) = §(type(T))`, that is the contract associated to the type.
+
+#### Records
+
+It remains to treat the case of records, that can mix previously type and term
+constructs.
+
+- a row declaration `foo : Num` is translated as it is. This is actually already
```

Comment:
What about `foo : Num = 1` ?

Reply:
I think this part has now more explanations about this case. Let me know if this is the case.

Reply:
Sorry, actually here we are inside the translation of terms. So `{foo : Num = 1}` is already a totally valid term, no issue.